### PR TITLE
Add ability to return a JSON body from an exception

### LINKED ISSUE #209: Allow for JSON Exception Bodies
I'm currently working on several projects that provide a JSON API using WebOb. Currently, however, whenever we use a `webob.exc` exception to return an error to the user (e.g., `webob.exc.HTTPBadRequest`) the body of that message is always in a content-type other than what they're expecting (HTML if they don't specify an Accept header, plain-text otherwise). There doesn't seem to be a pleasant, convenient, or simple way to make it use JSON beyond something like (the untested) following code:

```py
import string

import webob.exc


class WSGIHTTPException(webob.exc.WSGIHTTPException):
    body_template_obj = string.Template('{"code", ${status}, "message": "${body}", "title": "${title}"}'
    plain_template_obj = string.Template('{"error": ${body}}')


class HTTPBadRequest(webob.exc.HTTPBadRequest, WSGIHTTPException):
    pass


class HTTPUnauthored(webob.exc.HTTPBadRequest, WSGIHTTPException):
    pass


# etc.
```

This is particularly problematic because we have to redefine all of the exceptions we want to use to doubly inherit from our new sub-classed `WSGIHTTPException` and the original. It also doesn't handle the fact that we have to basically copy and paste [generate_response][] into our subclass so that we set the appropriate content-type header.

Is it too much to ask to either:

A) Add support for JSON response bodies in `WSGIHTTPException`s, or
B) Make `WSGIHTTPException` slightly more modular so we can only override parts we need?

[generate_response]: https://github.com/Pylons/webob/blob/7f98f694e7c1a569f53fb4085d084430ee8b2cc2/webob/exc.py#L302..L323

Thanks in advance,

### PR DESCRIPTION
**Pull request by sigmavirus24**
Closes #209

### CODE REVIEW

File: webob/exc.py [hunk 1]
```diff
@@ -308,6 +322,9 @@ def generate_response(self, environ, start_response):
         if accept and 'html' in accept or '*/*' in accept:
             content_type = 'text/html'
             body = self.html_body(environ)
+        elif accept and 'json' in accept:
```
Comment (bertjwregeer) [inline webob/exc.py]: If you have an XHR request that also still has `*/*` in it, this case will not get hit. Is this acceptable? Or do we want to see if JSON is listed first before checking for `HTML`/`*/*`?

**Comment by bertjwregeer** [comment]
If you could please make sure that coverage is back to 100% I will accept this PR.

**Comment by mmerickel** [comment]
I would really prefer a solution that parsed the accept and used `best_match` if we are going down this route of caring about accept.

```python
accept = webob.accept.Accept.parse(environ.get['HTTP_ACCEPT'])
match = accept.best_match('text/html', 'application/json')
...
```

**Comment by bertjwregeer** [comment]
We currently already care about `accept`, in that it checks if HTML exists or not. Using `best_match` would indeed be a better solution.

**Comment by bertjwregeer** [comment]
This kind of goes hand in hand with: https://github.com/Pylons/pyramid/pull/1378 in that we could use very similar to almost exactly the same code in both codebases since Pyramid's HTTPExceptions are a strict superset of the WebOb ones.

**Comment by sigmavirus24** [comment]
@bertjwregeer I've addressed the feedback thus far. Travis is running, but just wanted to let you know this should be ready for review again (I've 100% coverage locally.)

**Comment by mmerickel** [comment]
Do you have a need for passing `json_formatter` into `__init__`? That smells like feature creep to me.

**Comment by sigmavirus24** [comment]
@mmerickel it's mostly to give people a way to change the "template" for their JSON response but I guess they can just subclass and override that? I was trying to provide functionality similar to `body_template` and `text_template`. I don't know if that's desired though.

**Comment by sigmavirus24** [comment]
@mmerickel @bertjwregeer any further thoughts on `json_formatter` or this PR?

**Comment by bertjwregeer** [comment]
I personally don't mind adding the `json_formatter`. I think it might be a good idea that allows customization and would be helpful in my own projects so I don't have to sub-class anything.

**Comment by bertjwregeer** [comment]
Unless I hear a complaint I'll merge this in the next day or so.

**Comment by mmerickel** [comment]
LGTM, I just wanted to force someone to defend it before it goes in. Conforming to the analogous `text_template` etc is enough for me!

**Comment by sigmavirus24** [comment]
Yeah, we just can't use templates for JSON unfortunately. There are keywords to `json.dumps` that people may want to pass in the future too, but by default this should be enough. It might be fair, however, to ask people to subclass to pass those specifically. I don't need them, but I can imagine someone might, thoughts?

**Comment by bertjwregeer** [comment]
@sigmavirus24 If someone wants to pass particular keywords, then they can sub-class, otherwise we can keep going and ultimately WebOb will have full JSON parsing/validation/everything else ;-)

**Comment by sigmavirus24** [comment]
:ok_hand:

**Comment by sjmc7** [comment]
As an additional vote for this, it'd be really helpful for the openstack projects that use webob and generally work with JSON requests and responses.

**Comment by ttripp** [comment]
As an OpenStack developer, I can definitely agree that this would help keep us from having to re-implement this code.

**Comment by bertjwregeer** [comment]
As an OpenStack user, I don't want you guys re-implementing this code =)

**Comment by sigmavirus24** [comment]
Thanks @bertjwregeer! Let me know if I can help out with 1.6. It's clear a bunch of people would benefit from this. (I suspect @sjmc7 and @ttripp would also be willing to help out too.)

**Comment by bertjwregeer** [comment]
@sigmavirus24 appreciate that, I have tagged issues for 1.6 release with the milestone 1.6, they are visible here: https://github.com/Pylons/webob/milestones/Version%201.6

All of them are currently PR's that just need someone to verify A. that they fix the issue at end, and B. that they don't cause any backwards incompatible changes (and if it does, is there a way to do it in a BW-compat way?).

Even something as simple as pulling them down, verifying they work correctly, and then adding a +1 to the PR would be very helpful!

**Comment by sjmc7** [comment]
For what it's worth, i tested our service against the current master (bbca6e8ec1ae98501725817c315ca48aed5ea7fe) and everything seems to work as expected, though we don't use a lot of the features in that list. In particular, the change sigmavirus24 made seems to work as expected.

**Comment by sigmavirus24** [comment]
@sjmc7 can you pull the other changes in the [1.6 milestone](https://github.com/Pylons/webob/milestones/Version%201.6) and test your service against that all merged into master?

**Comment by sjmc7** [comment]
Yep, I pulled the current master, not just this patch.

**Comment by sigmavirus24** [comment]
@sjmc7 what about merging the pull requests in that milestone to master in your clone (this has already been merged). You could at least verify those don't seem to break backwards compataibility.

**Comment by sjmc7** [comment]
Ah, I'm sorry, i misunderstood. Been using gerrit for too long. I'll give them a try.

**Comment by sjmc7** [comment]
Ok, tested with all the patches in the 1.6 milestone and don't have any regressions. Thanks for setting me straight.

**Comment by bertjwregeer** [comment]
Awesome, good to know. I'm slowly going through them now and merging them, as well as doing the necessary work to document the changes.

**Comment by bertjwregeer** [comment]
@sjmc7 @sigmavirus24 Released 1.6.0a0, please test on projects and let me know if there are any issues :-).

Thanks for all the help guys!

**Comment by sjmc7** [comment]
Done some manual testing and run our tests on 1.6.0a0, and didn't find any problems (though as i say, we don't exercise a huge number of features). Thanks both of you for getting this done!

**Comment by sigmavirus24** [comment]
I'll try to run some other project's tests against 1.6.0a0 this weekend. Thanks @bertjwregeer

**Comment by sigmavirus24** [comment]
So apparently the decision in this PR that decides on returning HTML or Plain text has caused some minor backwards incompatibility. Running tests with [glance](/openstack/glance) and 1.6.0a0 showed that our (overly specific) tests were receiving plain text instead of HTML in some cases. (I haven't debugged further to determine why though.)

Also @flaper87 should be able to let us know if he saw any other test failures.

**Comment by bertjwregeer** [comment]
@sigmavirus24 Before I ship 1.6.0, are there any changes that need to happen, or is this something that needs to be fixed in the Glance tests?

**Comment by sigmavirus24** [comment]
Both? So I'm not sure 1.6.0 should ship with a (minor) backwards incompatibility although I still haven't looked at the failing test cases that @flaper87 found. I also think that Glance's tests (as I hinted at above) are a bit too ... specific and are testing things they shouldn't be.

**Comment by sigmavirus24** [comment]
That said, if you're content with 1.6.0 having what seems like it might be a minor backwards incompatibility, you could always fix that with 1.6.1 if anyone else runs into trouble with it.

**Comment by mmerickel** [comment]
I personally think it's a little silly to assume that webob can never change the format of the bundled exceptions. The library provides ways to override the format among other things if that's mission-critical. Most tools will ignore the content anyway and focus on the status code. Anyway it seems to me the change is very minor and acceptable.

**Comment by sigmavirus24** [comment]
@mmerickel changing the format is one thing. Returning a totally different content-type is another, I'd think. Again, I'm only arguing this because this was clearly the result of my change. I'm not going to stop y'all from releasing 1.6.0 though.

**Comment by bertjwregeer** [comment]
The content-type returned should be based upon what the original requestor asked for. If this is not the case, then things need to be changed, but if it is correct currently and the tests are failing because WebOb is now arguably doing the correct thing, then that's something we need to take into account.

**Comment by FlaPer87** [comment]
I didn't see other tests break other than the ones I reported. Those tests, as @sigmavirus24 mentioned are extremely specific and they could be changed. @bertjwregeer When is the release scheduled? I'd appreciate a couple of days for us to fix them before the release.

**Comment by bertjwregeer** [comment]
Is the behaviour that WebOb has correct? That's what I want to make sure of first and foremost. If you guys are expecting a different response and that is broken with the new code, why is that the case? Is the Accept matching not accurate?

**Comment by bertjwregeer** [comment]
@FlaPer87 @sigmavirus24 @sjmc7 Are you guys okay with the changes moving forward?

I noticed that the tests that have changed were expecting HTML, when their accept headers stated they wanted JSON (which this change fixes for the better). Are we good to ship this?

**Comment by FlaPer87** [comment]
@bertjwregeer We are, those tests have been fixed. Thanks a bunch for waiting on us and following-up! You just won a cake: :cake:

**Comment by sjmc7** [comment]
Yep, we're good with it too, thanks!

**Comment by sigmavirus24** [comment]
:shipit:

**Comment by bertjwregeer** [comment]
Cheers! Will ship soon.

**Comment by ttripp** [comment]
\O/

**Pull request merged**